scope/cluster.go

```diff
@@ -153,6 +161,44 @@ func (c *Cluster) IsDeleted() bool {
 	return !c.Cluster.DeletionTimestamp.IsZero() || !c.IonosCluster.DeletionTimestamp.IsZero()
 }
 
+func (c *Cluster) currentRequestByDatacenterLockKey() string {
+	return string(c.Cluster.UID) + "/currentRequestByDatacenter"
+}
+
+// GetCurrentRequestByDatacenter returns the current provisioning request for the given data center and a boolean
+// indicating if it exists.
+func (c *Cluster) GetCurrentRequestByDatacenter(datacenterID string) (infrav1.ProvisioningRequest, bool) {
+	lockKey := c.currentRequestByDatacenterLockKey()
+	_ = c.Locker.Lock(context.Background(), lockKey)
+	defer c.Locker.Unlock(lockKey)
+	req, ok := c.IonosCluster.Status.CurrentRequestByDatacenter[datacenterID]
+	return req, ok
+}
+
+// SetCurrentRequestByDatacenter sets the current provisioning request for the given data center.
+// This function makes sure that the map is initialized before setting the request.
+func (c *Cluster) SetCurrentRequestByDatacenter(datacenterID, method, status, requestPath string) {
+	lockKey := c.currentRequestByDatacenterLockKey()
+	_ = c.Locker.Lock(context.Background(), lockKey)
+	defer c.Locker.Unlock(lockKey)
+	if c.IonosCluster.Status.CurrentRequestByDatacenter == nil {
+		c.IonosCluster.Status.CurrentRequestByDatacenter = map[string]infrav1.ProvisioningRequest{}
+	}
+	c.IonosCluster.Status.CurrentRequestByDatacenter[datacenterID] = infrav1.ProvisioningRequest{
+		Method:      method,
+		RequestPath: requestPath,
+		State:       status,
+	}
+}
+
+// DeleteCurrentRequestByDatacenter deletes the current provisioning request for the given data center.
+func (c *Cluster) DeleteCurrentRequestByDatacenter(datacenterID string) {
+	lockKey := c.currentRequestByDatacenterLockKey()
+	_ = c.Locker.Lock(context.Background(), lockKey)
+	defer c.Locker.Unlock(lockKey)
```

Comment:
Super nit: `defer` causes a bit of overhead. As we do not panic or exit early we can also just Unlock after the `delete` call. But that is crying on a high level

Reply:
True, but I'd prefer it leave this as is: IMO it's easier for a reader looking for the lock-defer-unlock pattern. I find it easier than to look further down if calling `Unlock()` was forgotten.